File: config/funding_perp_spread_test.go
package config

import "testing"

func TestGenerateBotIDFundingPerpSpread_orderStable(t *testing.T) {
	fp := &FundingPerpSpreadConfig{
		LegA: FundingPerpLeg{Exchange: "okx", Symbol: "BTC-USDT-SWAP"},
		LegB: FundingPerpLeg{Exchange: "binance", Symbol: "BTCUSDT"},
	}
	fp2 := &FundingPerpSpreadConfig{
		LegA: FundingPerpLeg{Exchange: "binance", Symbol: "BTCUSDT"},
		LegB: FundingPerpLeg{Exchange: "okx", Symbol: "BTC-USDT-SWAP"},
	}
	if GenerateBotIDFundingPerpSpread(fp) != GenerateBotIDFundingPerpSpread(fp2) {
		t.Fatal("leg order should not change generated id")
	}
}

func TestValidateFundingPerpSpread_sameLeg(t *testing.T) {
	err := ValidateFundingPerpSpread(&FundingPerpSpreadConfig{
		LegA: FundingPerpLeg{Exchange: "binance", Symbol: "BTCUSDT"},
		LegB: FundingPerpLeg{Exchange: "binance", Symbol: "BTCUSDT"},
	})
	if err == nil {
		t.Fatal("expected error for identical legs")
	}
}

File: webui/src/utils/fundingPerpSpread.test.ts
import { describe, expect, it } from 'vitest'
import { parseFundingPerpSpread } from './fundingPerpSpread'
import type { BotDetailInfo } from '../services/api'

function baseBot(overrides: Partial<BotDetailInfo>): BotDetailInfo {
  return {
    bot_id: 'b1',
    name: 't',
    exchange: 'binance',
    symbol: 'BTCUSDT',
    market_type: 'futures',
    running: false,
    ...overrides,
  } as BotDetailInfo
}

describe('parseFundingPerpSpread', () => {
  it('returns null when market_type is not funding_perp_spread', () => {
    expect(parseFundingPerpSpread(baseBot({ market_type: 'futures' }))).toBeNull()
  })

  it('returns null when legs are incomplete', () => {
    expect(
      parseFundingPerpSpread(
        baseBot({
          market_type: 'funding_perp_spread',
          config: { funding_perp_spread: { leg_a: { exchange: 'a', symbol: 'x' }, leg_b: { exchange: '', symbol: 'y' } } },
        })
      )
    ).toBeNull()
  })

  it('returns config when valid', () => {
    const fp = {
      leg_a: { exchange: 'binance', symbol: 'BTCUSDT' },
      leg_b: { exchange: 'okx', symbol: 'BTC-USDT-SWAP' },
    }
    const r = parseFundingPerpSpread(
      baseBot({
        market_type: 'funding_perp_spread',
        config: { funding_perp_spread: fp },
      })
    )
    expect(r).toEqual(fp)
  })
})
